### DISCUSSION TITLE
Email verification link `self-service` suffix

### DISCUSSION THREAD

**Discussion by dro-sh** [ory/kratos#2730] [Q&A] [answered]
I would like to add email verification for users. But link on email has next format: `https://example.com/self-service/verification/?flow=<id>&token=<token>`.
I am using Next JS template https://github.com/ory/kratos-selfservice-ui-react-nextjs
I set `selfservice.flows.verification.ui_url` to `https://example.com/verification`
1. How remove suffix `self-service`?
2. What I should do when user go through link on email? Should I call [POST submit](https://www.ory.sh/docs/kratos/reference/api#operation/submitSelfServiceVerificationFlow) verification with flowID and token?

**Comment by dro-sh** [comment, marked as answer]
As I understand, verification URL should has host to Kratos. If my NextJS app on `https://example.com`, my Kratos Instance should be on `https://auth.example.com` then link with flowID and token will be sent to Kratos. Am I right?

**Reply by aeneasr** [reply]
Correct!

**Comment by armoleon** [comment]
Hi,

I'm testing email verification on my localhost and consuming kratos api via back-end app.
I also want to remove the self-service suffix from the link in the e-mail.

I tried followings:
1- set SELFSERVICE_FLOWS_VERIFICATION_UI_URL environment variable to http://127.0.0.1/verification
2- set kratos.yaml config file
selfservice:
  flows:
    verification:
      enabled: true
      ui_url: http://127.0.0.1/verification

But the self-service suffix in the link still remains. Is there any way to remove it?

**Reply by McKean** [reply]
@armoleon the url in the email should lead to kratos, which will then redirect to the UI.

Basically:
`SELFSERVICE_METHODS_[CODE|LINK]_CONFIG_BASE_URL` > `SELFSERVICE_FLOWS_VERIFICATION_UI_URL`
`https://mykratos/self-service/verification?....` > `https://myui/whatever?flow=...`

meaning self-service is required.

If for whatever reason you want to force your way, I recommend just overriding the email templates.

**Comment by dmitry-kostin** [comment]
I can say the verification URL is the final barrier to hiding all Kratos opinionated URLs behind my own proxy, which is quite frustrating. Sticking to the config values could be a more consistent and straightforward approach than replacing variable in our own emails, which I suspect could be tricky task again, as we have only {{ .VerificationURL }} variable there, according [to docs](https://www.ory.sh/docs/kratos/emails-sms/custom-email-templates)

**Reply by fraschm1998** [reply]
did you figure out a way around it?

**Reply by dmitry-kostin** [reply]
Nope 😀

**Reply by mtillberg** [reply]
I ran into this as well, but I found a way around it.  The VerificationURL is composed of either selfservice.methods.link.config.base_url or serve.public.base_url which then has "/self-service/verification" appended to it.  In the Kratos code, the various /self-service strings are defined in selfservice/flow/*/handler.go in a const block.  What you can do is override the templates with your own version, and use [sprig](https://masterminds.github.io/sprig/) functions to modify variables in the template.   I was able to set the template in the config:

```
courier:
  templates:
    verification_code:
      valid:
        email:
          body:
            html: file:///etc/config/kratos/templates/verification_code/valid/email.body.gotmpl
            plaintext: file:///etc/config/kratos/templates/verification_code/valid/email.body.gotmpl
```
Then, within the template modify VerificationURL using replace.  In my case, all I want to do is remove the "self-service" part of the URL:
```
<a href="{{ .VerificationURL | replace "/self-service" "" }}">{{ .VerificationURL | replace "/self-service" "" }}</a>
```
The built-in templates can be found in courier/template/courier/builtin/templates

**Reply by Tan-Aki** [reply]
Thank you @mtillberg !!

**Reply by elthariel** [reply]
For future reference. The `selfservice.methods.link.config.base_url` config key support was silently dropped, afaict from the code in https://github.com/ory/kratos/commit/db8a94ef538cfe0f1fc08bf5366643222a7bedfc
So there's no other method than dirty template tricks
